apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientImpl.SingleChannel.cs

```diff
@@ -392,13 +401,34 @@ private void HandlePullMessageData(Task<bool> moveNextTask)
                 receivedMessages.Clear();
                 // Get all ack-ids, used to extend leases as required.
                 var msgIds = new HashSet<string>(msgs.Select(x => x.AckId));
+                _receiptModAckForExactlyOnceDelivery = _exactlyOnceDeliveryEnabled;
+                if (_exactlyOnceDeliveryEnabled)
+                {
+                    // Populate receipt ModAck dictionary before the lease extends are send.
+                    PopulateAckIdsForReceiptModAck(msgIds);
+                }
+
                 // Send an initial "lease-extension"; which starts the server timer.
                 HandleExtendLease(msgIds, null);
-                // Asynchonously start message processing. Handles flow, and calls the user-supplied message handler.
-                // Uses Task.Run(), so not to clog up this "master" thread with per-message processing.
-                Task messagesTask = _taskHelper.Run(() => ProcessPullMessagesAsync(msgs, msgIds));
-                // Once all received messages have been queued for processing, read the stream for more messages.
-                Add(messagesTask, Next(true, () => HandlePullMoveNext(null)));
+                if (_exactlyOnceDeliveryEnabled)
+                {
+                    // In the exactly-once delivery flow, we must send only those messages to the user handler which had successful receipt ModAck responses.
+                    // This is necessary to prevent message redelivery. We get the status of success or permanent failure instantly.
```

Comment:
> We get the status of success or permanent failure instantly.

I don't understand this comment. How can we get success or permanent failure instantly - we're waiting for an RPC, right?